File: android/app/src/main/java/dev/kian/gymapp/MaisGenieXRuntimePlugin.java
package dev.kian.gymapp;

import android.os.Debug;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.Iterator;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicLong;

@CapacitorPlugin(name = "MaisGenieXRuntime")
public final class MaisGenieXRuntimePlugin extends Plugin {
    private static final String MODEL_ID = "qwen.qwen3-4b";
    private static final String RUNTIME_VERSION = "QAIRT 2.45.0.260326154327";
    private static final int CONTEXT_TOKENS = 12_288;
    private static final int READ_BUFFER_BYTES = 64 * 1024;
    private static final ExecutorService EXECUTOR = Executors.newSingleThreadExecutor();
    private static final AtomicBoolean RUNNING = new AtomicBoolean(false);
    private static final AtomicLong ACTIVE_HANDLE = new AtomicLong(0L);
    private static final List<String> REQUIRED_MODEL_FILES = Arrays.asList(
        "added_tokens.json",
        "config.json",
        "genie_config.json",
        "htp_backend_ext_config.json",
        "merges.txt",
        "metadata.json",
        "part1_of_4.bin",
        "part2_of_4.bin",
        "part3_of_4.bin",
        "part4_of_4.bin",
        "sample_prompt.txt",
        "special_tokens_map.json",
        "tokenizer.json",
        "tokenizer_config.json",
        "vocab.json"
    );

    @PluginMethod
    public void getStatus(PluginCall call) {
        try {
            call.resolve(status());
        } catch (Exception error) {
            call.reject(rootMessage(error), error);
        }
    }

    @PluginMethod
    public void runBaseline(PluginCall call) {
        final String modelId;
        final String prompt;
        final String systemInstruction;
        try {
            modelId = required(call, "modelId");
            if (!MODEL_ID.equals(modelId)) {
                throw new IllegalArgumentException("The GenieX adapter currently supports Qwen3-4B only.");
            }
            prompt = optional(
                call,
                "prompt",
                "Reply with exactly two short sentences. First confirm that this response was generated locally. Second state that verified evidence should be preferred over an unbounded transcript."
            );
            systemInstruction = optional(
                call,
                "systemInstruction",
                "You are the bounded native Qwen runtime check for My Mettle. Be concise, factual and never claim access to data that was not supplied."
            );
            ensureReady();
            if (!RUNNING.compareAndSet(false, true)) {
                throw new IllegalStateException("A GenieX run is already active.");
            }
        } catch (Exception error) {
            call.reject(rootMessage(error), error);
            return;
        }
        EXECUTOR.execute(() -> executeBaseline(call, modelId, prompt, systemInstruction));
    }

    @PluginMethod
    public void cancelRun(PluginCall call) {
        JSObject result = new JSObject();
        result.put("requested", false);
        result.put("supported", false);
        result.put(
            "reason",
            "The currently integrated Genie C API does not expose a validated cancellation ABI. The bounded probe must finish before another run starts."
        );
        call.resolve(result);
    }

    private void executeBaseline(PluginCall call, String modelId, String prompt, String systemInstruction) {
        final long startedAt = System.currentTimeMillis();
        final long memoryBefore = currentPssBytes();
        long peakPss = memoryBefore;
        long handle = 0L;
        long loadMs = 0L;
        long firstChunkMs = -1L;
        long generationMs = 0L;
        long unloadMs = 0L;
        String rawOutput = "";
        String finalOutput = "";
        String profileJson = "";
        String state = "completed";
        String failure = null;

        emitProgress(modelId, "loading", 0L, 0);
        try {
            MaisGenieXNative.requireLoaded();
            String configJson = buildRuntimeConfig();
            long loadStarted = System.currentTimeMillis();
            handle = MaisGenieXNative.nativeCreate(
                MaisQairtRuntimePlugin.arm64Directory(getContext().getFilesDir()).getAbsolutePath(),
                MaisQairtRuntimePlugin.hexagonDirectory(getContext().getFilesDir()).getAbsolutePath(),
                configJson
            );
            if (handle == 0L) throw new IllegalStateException("Genie returned a null runtime handle.");
            ACTIVE_HANDLE.set(handle);
            loadMs = System.currentTimeMillis() - loadStarted;
            peakPss = Math.max(peakPss, currentPssBytes());

            emitProgress(modelId, "generating", loadMs, 0);
            rawOutput = nullToEmpty(
                MaisGenieXNative.nativeQuery(handle, buildTaggedPrompt(systemInstruction, prompt))
            ).trim();
            generationMs = Math.max(0L, MaisGenieXNative.nativeGetLastQueryMs(handle));
            firstChunkMs = MaisGenieXNative.nativeGetLastFirstChunkMs(handle);
            profileJson = nullToEmpty(MaisGenieXNative.nativeGetLastProfileJson(handle));
            finalOutput = extractFinalOutput(rawOutput);
            if (finalOutput.isEmpty()) {
                throw new IllegalStateException("Qwen completed without returning a final response.");
            }
            peakPss = Math.max(peakPss, currentPssBytes());
        } catch (Throwable error) {
            state = "failed";
            failure = rootMessage(error);
        } finally {
            long unloadStarted = System.currentTimeMillis();
            if (handle != 0L) {
                try {
                    MaisGenieXNative.nativeFree(handle);
                } catch (Throwable closeError) {
                    if (failure == null) {
                        state = "failed";
                        failure = "Qwen generated output but failed to unload cleanly: " + rootMessage(closeError);
                    }
                }
            }
            ACTIVE_HANDLE.set(0L);
            unloadMs = System.currentTimeMillis() - unloadStarted;
            peakPss = Math.max(peakPss, currentPssBytes());
            RUNNING.set(false);
        }

        long completedAt = System.currentTimeMillis();
        try {
            JSObject result = new JSObject();
            result.put("state", state);
            result.put("success", "completed".equals(state));
            result.put("modelId", modelId);
            result.put("runtime", "geniex-qairt");
            result.put("runtimeVersion", RUNTIME_VERSION);
            result.put("backend", "npu");
            result.put("contextTokens", CONTEXT_TOKENS);
            result.put("startedAtEpochMs", startedAt);
            result.put("completedAtEpochMs", completedAt);
            result.put("loadMs", loadMs);
            result.put("firstChunkLatencyMs", firstChunkMs);
            result.put("generationMs", generationMs);
            result.put("unloadMs", unloadMs);
            result.put("totalMs", completedAt - startedAt);
            result.put("memoryBeforeBytes", memoryBefore);
            result.put("peakPssBytes", peakPss);
            result.put("memoryAfterBytes", currentPssBytes());
            result.put("rawOutput", rawOutput);
            result.put("finalOutput", finalOutput);
            result.put("outputChars", finalOutput.length());
            result.put("profile", profileMetrics(profileJson));
            result.put("error", failure == null ? JSONObject.NULL : failure);
            writeLastResult(result);
            emitProgress(modelId, state, loadMs, finalOutput.length());
            call.resolve(result);
        } catch (Exception error) {
            call.reject(rootMessage(error), error);
        }
    }

    private JSObject status() throws Exception {
        JSONArray missingFiles = missingModelFiles();
        File arm64 = MaisQairtRuntimePlugin.arm64Directory(getContext().getFilesDir());
        File hexagon = MaisQairtRuntimePlugin.hexagonDirectory(getContext().getFilesDir());
        boolean runtimeInstalled = new File(arm64, "libGenie.so").isFile()
            && new File(arm64, "libQnnSystem.so").isFile()
            && new File(arm64, "libQnnHtp.so").isFile()
            && new File(arm64, "libQnnHtpPrepare.so").isFile()
            && hasLibrary(arm64, "Stub.so")
            && hasLibrary(hexagon, "Skel.so");

        JSObject result = new JSObject();
        result.put("ready", missingFiles.length() == 0 && runtimeInstalled && MaisGenieXNative.isLoaded());
        result.put("running", RUNNING.get());
        result.put("modelId", MODEL_ID);
        result.put("runtime", "geniex-qairt");
        result.put("runtimeVersion", RUNTIME_VERSION);
        result.put("backend", "npu");
        result.put("contextTokens", CONTEXT_TOKENS);
        result.put("bundleReady", missingFiles.length() == 0);
        result.put("missingModelFiles", missingFiles);
        result.put("runtimeInstalled", runtimeInstalled);
        result.put("bridgeLoaded", MaisGenieXNative.isLoaded());
        result.put(
            "bridgeError",
            MaisGenieXNative.loadError() == null ? JSONObject.NULL : MaisGenieXNative.loadError()
        );
        result.put("lastResult", readLastResult());
        return result;
    }

    private void ensureReady() throws Exception {
        JSObject current = status();
        if (!current.getBool("bundleReady")) {
            throw new IllegalStateException("The complete verified Qwen3-4B 12K model pack is not installed.");
        }
        if (!current.getBool("runtimeInstalled")) {
            throw new IllegalStateException("Import the matching QAIRT 2.45 Android runtime before running Qwen.");
        }
        MaisGenieXNative.requireLoaded();
    }

    private String buildRuntimeConfig() throws Exception {
        File models = modelDirectory();
        JSONObject config = new JSONObject(readText(new File(models, "genie_config.json")));
        JSONObject dialog = config.getJSONObject("dialog");
        dialog.getJSONObject("tokenizer").put(
            "path",
            new File(models, "tokenizer.json").getAbsolutePath()
        );
        JSONObject engine = dialog.getJSONObject("engine");
        engine.getJSONObject("backend").put(
            "extensions",
            new File(models, "htp_backend_ext_config.json").getAbsolutePath()
        );
        JSONArray binaries = engine
            .getJSONObject("model")
            .getJSONObject("binary")
            .getJSONArray("ctx-bins");
        for (int index = 0; index < binaries.length(); index += 1) {
            String name = new File(binaries.getString(index)).getName();
            File binary = new File(models, name);
            if (!binary.isFile()) {
                throw new IllegalStateException("Qwen context binary is missing: " + name + ".");
            }
            binaries.put(index, binary.getAbsolutePath());
        }
        return config.toString();
    }

    private String buildTaggedPrompt(String systemInstruction, String userPrompt) {
        String tagged;
        try {
            JSONObject metadata = new JSONObject(readText(new File(modelDirectory(), "metadata.json")));
            JSONObject template = metadata.getJSONObject("genie").getJSONObject("chat_template");
            tagged = template.getString("system_prefix")
                + systemInstruction
                + template.getString("system_suffix")
                + template.getString("user_prefix")
                + userPrompt
                + template.getString("user_suffix")
                + template.getString("assistant_prefix");
        } catch (Exception ignored) {
            tagged = "<|im_start|>system\n" + systemInstruction + "<|im_end|>\n"
                + "<|im_start|>user\n" + userPrompt + "<|im_end|>\n"
                + "<|im_start|>assistant\n";
        }
        return tagged + "<think>\n\n</think>\n";
    }

    private JSObject profileMetrics(String profileJson) {
        JSObject metrics = new JSObject();
        metrics.put("available", !profileJson.trim().isEmpty());
        metrics.put("timeToFirstTokenMs", JSONObject.NULL);
        metrics.put("tokenGenerationRate", JSONObject.NULL);
        metrics.put("promptProcessingRate", JSONObject.NULL);
        metrics.put("promptTokens", JSONObject.NULL);
        metrics.put("generatedTokens", JSONObject.NULL);
        if (profileJson.trim().isEmpty()) return metrics;
        try {
            JSONObject profile = new JSONObject(profileJson);
            findMetric(profile, "time-to-first-token", metrics, "timeToFirstTokenMs");
            findMetric(profile, "token-generation-rate", metrics, "tokenGenerationRate");
            findMetric(profile, "prompt-processing-rate", metrics, "promptProcessingRate");
            findMetric(profile, "num-prompt-tokens", metrics, "promptTokens");
            findMetric(profile, "num-generated-tokens", metrics, "generatedTokens");
        } catch (Exception error) {
            metrics.put("parseError", rootMessage(error));
        }
        return metrics;
    }

    private boolean findMetric(
        Object node,
        String targetKey,
        JSObject output,
        String outputKey
    ) throws Exception {
        if (node instanceof JSONObject) {
            JSONObject object = (JSONObject) node;
            if (object.has(targetKey)) {
                Object value = object.get(targetKey);
                if (value instanceof JSONObject && ((JSONObject) value).has("value")) {
                    JSONObject wrapped = (JSONObject) value;
                    output.put(outputKey, wrapped.get("value"));
                    if (wrapped.has("unit")) output.put(outputKey + "Unit", wrapped.get("unit"));
                } else {
                    output.put(outputKey, value);
                }
                return true;
            }
            Iterator<String> keys = object.keys();
            while (keys.hasNext()) {
                String key = keys.next();
                if (findMetric(object.get(key), targetKey, output, outputKey)) return true;
            }
        } else if (node instanceof JSONArray) {
            JSONArray array = (JSONArray) node;
            for (int index = 0; index < array.length(); index += 1) {
                if (findMetric(array.get(index), targetKey, output, outputKey)) return true;
            }
        }
        return false;
    }

    private String extractFinalOutput(String raw) {
        String output = nullToEmpty(raw).trim();
        int closingThink = output.lastIndexOf("</think>");
        if (closingThink >= 0) output = output.substring(closingThink + "</think>".length()).trim();
        return output.replace("<|im_end|>", "").trim();
    }

    private JSONArray missingModelFiles() {
        JSONArray missing = new JSONArray();
        File directory = modelDirectory();
        for (String name : REQUIRED_MODEL_FILES) {
            if (!new File(directory, name).isFile()) missing.put(name);
        }
        return missing;
    }

    private boolean hasLibrary(File directory, String suffix) {
        File[] matches = directory.listFiles(
            (parent, name) -> name.startsWith("libQnnHtpV") && name.endsWith(suffix)
        );
        return matches != null && matches.length > 0;
    }

    private void emitProgress(String modelId, String state, long loadMs, int outputChars) {
        JSObject event = new JSObject();
        event.put("modelId", modelId);
        event.put("state", state);
        event.put("backend", "npu");
        event.put("loadMs", loadMs);
        event.put("outputChars", outputChars);
        event.put("capturedAtEpochMs", System.currentTimeMillis());
        notifyListeners("genieXInferenceProgress", event);
    }

    private long currentPssBytes() {
        return Debug.getPss() * 1024L;
    }

    private File modelDirectory() {
        return new File(getContext().getFilesDir(), "mais-models");
    }

    private File runtimeResultDirectory() {
        return new File(getContext().getFilesDir(), "mais-runtime");
    }

    private File lastResultFile() {
        return new File(runtimeResultDirectory(), "last-qwen-qwen3-4b.json");
    }

    private void writeLastResult(JSObject result) throws Exception {
        File directory = runtimeResultDirectory();
        if (!directory.exists() && !directory.mkdirs()) {
            throw new IllegalStateException("Could not create the MAIS runtime directory.");
        }
        try (BufferedOutputStream output = new BufferedOutputStream(new FileOutputStream(lastResultFile()))) {
            output.write(result.toString(2).getBytes(StandardCharsets.UTF_8));
        }
    }

    private Object readLastResult() {
        try {
            if (!lastResultFile().isFile()) return JSONObject.NULL;
            return new JSONObject(readText(lastResultFile()));
        } catch (Exception ignored) {
            return JSONObject.NULL;
        }
    }

    private String readText(File file) throws Exception {
        try (
            BufferedInputStream input = new BufferedInputStream(new FileInputStream(file), READ_BUFFER_BYTES);
            ByteArrayOutputStream output = new ByteArrayOutputStream()
        ) {
            byte[] buffer = new byte[READ_BUFFER_BYTES];
            int count;
            while ((count = input.read(buffer)) >= 0) {
                if (count > 0) output.write(buffer, 0, count);
            }
            return output.toString(StandardCharsets.UTF_8.name());
        }
    }

    private String required(PluginCall call, String key) {
        String value = call.getString(key);
        if (value == null || value.trim().isEmpty()) {
            throw new IllegalArgumentException(key + " is required.");
        }
        return value.trim();
    }

    private String optional(PluginCall call, String key, String fallback) {
        String value = call.getString(key, fallback);
        if (value == null || value.trim().isEmpty()) return fallback;
        return value.trim();
    }

    private String nullToEmpty(String value) {
        return value == null ? "" : value;
    }

    private String rootMessage(Throwable error) {
        Throwable current = error;
        while (current.getCause() != null) current = current.getCause();
        String message = current.getMessage();
        return message == null || message.trim().isEmpty()
            ? current.getClass().getSimpleName()
            : message;
    }

    @Override
    protected void handleOnDestroy() {
        long handle = ACTIVE_HANDLE.getAndSet(0L);
        if (handle != 0L && !RUNNING.get()) {
            try {
                MaisGenieXNative.nativeFree(handle);
            } catch (Throwable ignored) {
                // Android reclaims the process native address space during teardown.
            }
        }
        super.handleOnDestroy();
    }
}

File: android/app/src/main/java/dev/kian/gymapp/MaisGenieXRuntimePlugin.kt
package dev.kian.gymapp

import android.os.Debug
import com.geniex.sdk.GenieXSdk
import com.geniex.sdk.LlmWrapper
import com.geniex.sdk.bean.ChatMessage
import com.geniex.sdk.bean.ComputeUnitValue
import com.geniex.sdk.bean.GenerationConfig
import com.geniex.sdk.bean.LlmCreateInput
import com.geniex.sdk.bean.LlmStreamResult
import com.geniex.sdk.bean.ModelConfig
import com.geniex.sdk.bean.ProfilingData
import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch
import org.json.JSONObject
import java.io.File
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicReference
import kotlin.math.roundToLong

@CapacitorPlugin(name = "MaisGenieXRuntime")
class MaisGenieXRuntimePlugin : Plugin() {
    companion object {
        private const val MODEL_ID = "qwen.qwen3-4b"
        private const val MODEL_NAME = "Qwen3-4B"
        private const val GENIEX_ANDROID_VERSION = "0.3.5"
        private const val CONTEXT_TOKENS = 12_288
        private const val MAX_OUTPUT_TOKENS = 2_048
        private val REQUIRED_MODEL_FILES = listOf(
            "added_tokens.json",
            "config.json",
            "genie_config.json",
            "htp_backend_ext_config.json",
            "merges.txt",
            "metadata.json",
            "part1_of_4.bin",
            "part2_of_4.bin",
            "part3_of_4.bin",
            "part4_of_4.bin",
            "sample_prompt.txt",
            "special_tokens_map.json",
            "tokenizer.json",
            "tokenizer_config.json",
            "vocab.json",
        )
    }

    private val runtimeScope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val running = AtomicBoolean(false)
    private val cancelRequested = AtomicBoolean(false)
    private val activeWrapper = AtomicReference<LlmWrapper?>(null)
    private val sdkLock = Any()

    @Volatile
    private var sdkReady = false

    @Volatile
    private var sdkError: String? = null

    @PluginMethod
    fun getStatus(call: PluginCall) {
        try {
            call.resolve(status())
        } catch (error: Throwable) {
            call.reject(rootMessage(error), error)
        }
    }

    @PluginMethod
    fun runBaseline(call: PluginCall) {
        val modelId = call.getString("modelId")?.trim().orEmpty()
        if (modelId != MODEL_ID) {
            call.reject("The GenieX adapter currently supports Qwen3-4B only.")
            return
        }
        if (!ensureSdkReady()) {
            call.reject(sdkError ?: "The bundled GenieX runtime could not initialise.")
            return
        }
        val missing = missingModelFiles()
        if (missing.isNotEmpty()) {
            call.reject("The Qwen3-4B 12K pack is incomplete: ${missing.joinToString()}.")
            return
        }
        if (!running.compareAndSet(false, true)) {
            call.reject("A GenieX run is already active.")
            return
        }

        cancelRequested.set(false)
        val prompt = call.getString("prompt")?.trim().takeUnless { it.isNullOrEmpty() }
            ?: "Reason carefully, then reply with exactly two short sentences. First confirm that this response was generated locally. Second state that verified evidence should be preferred over an unbounded transcript."
        val systemInstruction = call.getString("systemInstruction")?.trim().takeUnless { it.isNullOrEmpty() }
            ?: "You are the bounded native Qwen runtime check for My Mettle. Use thinking mode before answering. Keep the final response concise, factual and limited to supplied evidence."

        runtimeScope.launch {
            executeBaseline(call, modelId, prompt, systemInstruction)
        }
    }

    @PluginMethod
    fun cancelRun(call: PluginCall) {
        val wrapper = activeWrapper.get()
        if (!running.get() || wrapper == null) {
            val result = JSObject()
            result.put("requested", false)
            result.put("supported", true)
            result.put("reason", "No GenieX generation is active.")
            call.resolve(result)
            return
        }

        cancelRequested.set(true)
        runtimeScope.launch {
            val stopResult = wrapper.stopStream()
            val result = JSObject()
            result.put("requested", stopResult.isSuccess)
            result.put("supported", true)
            result.put("reason", stopResult.exceptionOrNull()?.let(::rootMessage) ?: JSONObject.NULL)
            call.resolve(result)
        }
    }

    private suspend fun executeBaseline(
        call: PluginCall,
        modelId: String,
        prompt: String,
        systemInstruction: String,
    ) {
        val startedAt = System.currentTimeMillis()
        val memoryBefore = currentPssBytes()
        var peakPss = memoryBefore
        var loadMs = 0L
        var firstChunkMs = -1L
        var generationMs = 0L
        var unloadMs = 0L
        var thinkingCharacters = 0
        var finalOutput = ""
        var profile: ProfilingData? = null
        var state = "completed"
        var failure: String? = null
        var wrapper: LlmWrapper? = null

        emitProgress(modelId, "loading", 0L, 0)
        try {
            val modelDirectory = modelDirectory()
            val tokenizer = File(modelDirectory, "tokenizer.json")
            val loadStarted = System.currentTimeMillis()
            wrapper = LlmWrapper
                .builder()
                .llmCreateInput(
                    LlmCreateInput(
                        model_name = MODEL_NAME,
                        model_path = modelDirectory.absolutePath,
                        tokenizer_path = tokenizer.absolutePath,
                        config = ModelConfig(
                            nCtx = 0,
                            nGpuLayers = 0,
                            max_tokens = MAX_OUTPUT_TOKENS,
                            enable_thinking = true,
                        ),
                        runtime_id = GenieXSdk.PLUGIN_ID_QAIRT,
                        compute_unit = ComputeUnitValue.NPU.value,
                    ),
                )
                .build()
                .getOrThrow()
            activeWrapper.set(wrapper)
            loadMs = System.currentTimeMillis() - loadStarted
            peakPss = maxOf(peakPss, currentPssBytes())

            val template = wrapper
                .applyChatTemplate(
                    arrayOf(
                        ChatMessage("system", systemInstruction),
                        ChatMessage("user", prompt),
                    ),
                    tools = null,
                    enableThinking = true,
                )
                .getOrThrow()

            emitProgress(modelId, "generating", loadMs, 0)
            val output = StringBuilder()
            val generationStarted = System.currentTimeMillis()
            wrapper.generateStreamFlow(
                template.formattedText,
                GenerationConfig(maxTokens = MAX_OUTPUT_TOKENS),
            ).collect { streamResult ->
                when (streamResult) {
                    is LlmStreamResult.Token -> {
                        if (firstChunkMs < 0 && streamResult.text.isNotEmpty()) {
                            firstChunkMs = System.currentTimeMillis() - generationStarted
                        }
                        output.append(streamResult.text)
                        emitProgress(modelId, "generating", loadMs, output.length)
                    }
                    is LlmStreamResult.Completed -> profile = streamResult.profile
                    is LlmStreamResult.Error -> throw streamResult.throwable
                }
            }
            generationMs = System.currentTimeMillis() - generationStarted

            if (cancelRequested.get()) {
                state = "cancelled"
            } else {
                val rawOutput = output.toString().trim()
                thinkingCharacters = countThinkingCharacters(rawOutput)
                if (!hasCompletedThinkingSection(rawOutput) || thinkingCharacters <= 0) {
                    throw IllegalStateException(
                        "Qwen did not complete a thinking section before its final response. The output budget may have ended during reasoning.",
                    )
                }
                finalOutput = extractFinalOutput(rawOutput)
                if (finalOutput.isEmpty()) {
                    throw IllegalStateException("Qwen completed thinking without returning a final response.")
                }
            }
            peakPss = maxOf(peakPss, currentPssBytes())
        } catch (error: Throwable) {
            if (cancelRequested.get()) {
                state = "cancelled"
            } else {
                state = "failed"
                failure = rootMessage(error)
            }
        } finally {
            val unloadStarted = System.currentTimeMillis()
            activeWrapper.set(null)
            if (wrapper != null) {
                runCatching { wrapper.stopStream() }
                runCatching { wrapper.destroy() }
                    .onFailure { closeError ->
                        if (failure == null && state == "completed") {
                            state = "failed"
                            failure = "Qwen generated output but failed to unload cleanly: ${rootMessage(closeError)}"
                        }
                    }
            }
            unloadMs = System.currentTimeMillis() - unloadStarted
            peakPss = maxOf(peakPss, currentPssBytes())
            running.set(false)
            cancelRequested.set(false)
        }

        val completedAt = System.currentTimeMillis()
        val resolvedFirstChunk = when {
            profile != null && profile!!.ttftMs >= 0.0 -> profile!!.ttftMs.roundToLong()
            else -> firstChunkMs
        }
        val result = JSObject()
        result.put("state", state)
        result.put("success", state == "completed")
        result.put("modelId", modelId)
        result.put("runtime", "geniex-qairt")
        result.put("runtimeVersion", runtimeVersion())
        result.put("sdkVersion", GENIEX_ANDROID_VERSION)
        result.put("distribution", "maven-central")
        result.put("backend", "npu")
        result.put("contextTokens", CONTEXT_TOKENS)
        result.put("thinkingEnabled", true)
        result.put("thinkingObserved", thinkingCharacters > 0)
        result.put("thinkingCharacters", thinkingCharacters)
        result.put("reasoningContentStored", false)
        result.put("startedAtEpochMs", startedAt)
        result.put("completedAtEpochMs", completedAt)
        result.put("loadMs", loadMs)
        result.put("firstChunkLatencyMs", resolvedFirstChunk)
        result.put("generationMs", generationMs)
        result.put("unloadMs", unloadMs)
        result.put("totalMs", completedAt - startedAt)
        result.put("memoryBeforeBytes", memoryBefore)
        result.put("peakPssBytes", peakPss)
        result.put("memoryAfterBytes", currentPssBytes())
        result.put("finalOutput", finalOutput)
        result.put("outputChars", finalOutput.length)
        result.put("profile", profileMetrics(profile))
        result.put("error", failure ?: JSONObject.NULL)
        writeLastResult(result)
        emitProgress(modelId, state, loadMs, finalOutput.length)
        call.resolve(result)
    }

    private fun status(): JSObject {
        val missing = missingModelFiles()
        val ready = ensureSdkReady()
        val result = JSObject()
        result.put("ready", missing.isEmpty() && ready)
        result.put("running", running.get())
        result.put("modelId", MODEL_ID)
        result.put("runtime", "geniex-qairt")
        result.put("runtimeVersion", runtimeVersion())
        result.put("sdkVersion", GENIEX_ANDROID_VERSION)
        result.put("distribution", "maven-central")
        result.put("backend", "npu")
        result.put("contextTokens", CONTEXT_TOKENS)
        result.put("thinkingEnabled", true)
        result.put("bundleReady", missing.isEmpty())
        result.put("missingModelFiles", JSArray(missing))
        result.put("runtimeInstalled", true)
        result.put("bridgeLoaded", ready)
        result.put("bridgeError", sdkError ?: JSONObject.NULL)
        result.put("lastResult", readLastResult())
        return result
    }

    private fun ensureSdkReady(): Boolean = synchronized(sdkLock) {
        if (sdkReady) return@synchronized true
        sdkError = null
        GenieXSdk.getInstance().init(
            context,
            object : GenieXSdk.InitCallback {
                override fun onSuccess() {
                    sdkReady = true
                    sdkError = null
                }

                override fun onFailure(reason: String) {
                    sdkReady = false
                    sdkError = reason.trim().ifEmpty { "GenieX initialisation failed." }
                }
            },
        )
        sdkReady
    }

    private fun runtimeVersion(): String {
        if (!sdkReady) return "GenieX Android $GENIEX_ANDROID_VERSION"
        return runCatching {
            GenieXSdk.getInstance().getPluginVersion(GenieXSdk.PLUGIN_ID_QAIRT)
        }.getOrNull()?.takeIf { it.isNotBlank() }
            ?: "GenieX Android $GENIEX_ANDROID_VERSION"
    }

    private fun profileMetrics(profile: ProfilingData?): JSObject {
        val metrics = JSObject()
        metrics.put("available", profile != null)
        metrics.put("timeToFirstTokenMs", profile?.ttftMs ?: JSONObject.NULL)
        metrics.put("promptProcessingTimeMs", profile?.promptTimeMs ?: JSONObject.NULL)
        metrics.put("decodeTimeMs", profile?.decodeTimeMs ?: JSONObject.NULL)
        metrics.put("tokenGenerationRate", profile?.decodingSpeed ?: JSONObject.NULL)
        metrics.put("tokenGenerationRateUnit", if (profile == null) JSONObject.NULL else "tok/s")
        metrics.put("promptProcessingRate", profile?.prefillSpeed ?: JSONObject.NULL)
        metrics.put("promptProcessingRateUnit", if (profile == null) JSONObject.NULL else "tok/s")
        metrics.put("promptTokens", profile?.promptTokens ?: JSONObject.NULL)
        metrics.put("generatedTokens", profile?.generatedTokens ?: JSONObject.NULL)
        metrics.put("stopReason", profile?.stopReason ?: JSONObject.NULL)
        return metrics
    }

    private fun missingModelFiles(): List<String> {
        val directory = modelDirectory()
        return REQUIRED_MODEL_FILES.filter { !File(directory, it).isFile }
    }

    private fun modelDirectory(): File = File(context.filesDir, "mais-models")

    private fun hasCompletedThinkingSection(raw: String): Boolean = raw.lastIndexOf("</think>") >= 0

    private fun countThinkingCharacters(raw: String): Int {
        val closing = raw.lastIndexOf("</think>")
        if (closing < 0) return 0
        val opening = raw.indexOf("<think>")
        val start = if (opening >= 0) opening + "<think>".length else 0
        return (closing - start).coerceAtLeast(0)
    }

    private fun extractFinalOutput(raw: String): String {
        val closing = raw.lastIndexOf("</think>")
        if (closing < 0) return ""
        return raw
            .substring(closing + "</think>".length)
            .replace("<|im_end|>", "")
            .trim()
    }

    private fun emitProgress(modelId: String, state: String, loadMs: Long, outputChars: Int) {
        val event = JSObject()
        event.put("modelId", modelId)
        event.put("state", state)
        event.put("backend", "npu")
        event.put("loadMs", loadMs)
        event.put("outputChars", outputChars)
        event.put("capturedAtEpochMs", System.currentTimeMillis())
        notifyListeners("genieXInferenceProgress", event)
    }

    private fun currentPssBytes(): Long = Debug.getPss() * 1024L

    private fun runtimeResultDirectory(): File = File(context.filesDir, "mais-runtime")

    private fun lastResultFile(): File = File(runtimeResultDirectory(), "last-qwen-qwen3-4b.json")

    private fun writeLastResult(result: JSObject) {
        val directory = runtimeResultDirectory()
        if (!directory.exists() && !directory.mkdirs()) {
            throw IllegalStateException("Could not create the MAIS runtime directory.")
        }
        lastResultFile().writeText(result.toString(2), Charsets.UTF_8)
    }

    private fun readLastResult(): Any {
        return runCatching {
            if (!lastResultFile().isFile) return@runCatching JSONObject.NULL
            JSONObject(lastResultFile().readText(Charsets.UTF_8)).apply {
                remove("rawOutput")
                put("reasoningContentStored", false)
            }
        }.getOrDefault(JSONObject.NULL)
    }

    private fun rootMessage(error: Throwable): String {
        var current = error
        while (current.cause != null) current = current.cause!!
        return current.message?.takeIf { it.isNotBlank() } ?: current.javaClass.simpleName
    }
}
